### ISSUE TITLE
Bogus null checks in MeterListener.SetMeasurementEventCallback<T>

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/runtime#68025]
### Description

If the `public void SetMeasurementEventCallback<T>(MeasurementCallback<T>? measurementCallback)` method of  System.Diagnostics.Metrics.MeterListener is called with a null callback, then it throws an InvalidOperationException that claims that the type is unsupported, even if MeterListener actually supports the type.

### Reproduction Steps

```C#
using System.Diagnostics.Metrics;

using var listener = new MeterListener();
listener.SetMeasurementEventCallback<int>(null);
```

### Expected behavior

`listener.SetMeasurementEventCallback<int>(null);` should discard any previously set callback for `int` and succeed.

### Actual behavior

Self-contradictory exception:

```
Unhandled exception. System.InvalidOperationException: System.Int32 is unsupported type for this operation. The only supported types are byte, short, int, long, float, double, and decimal.
   at System.Diagnostics.Metrics.MeterListener.SetMeasurementEventCallback[T](MeasurementCallback`1 measurementCallback)
   at Program.<Main>$(String[] args)
```

### Regression?

Not a regression; the original implementation in <https://github.com/dotnet/runtime/pull/52685> already did this.

### Known Workarounds

Provide a no-op callback instead of `null`.

### Configuration

.NET SDK 6.0.2, runtime 6.0.202. Windows 10 Version 21H2 on x64. 
It is not specific to this configuration.

### Other information

All the `measurementCallback is null` checks in MeterListener.SetMeasurementEventCallback are currently redundant because they are only evaluated if a condition like `measurementCallback is MeasurementCallback<byte> byteCallback` matches, which can't happen if `measurementCallback` is null. Perhaps the method should instead check `typeof(T) == typeof(byte)` etc. and then cast `(MeasurementCallback<byte>)(object)measurementCallback`.

<https://github.com/dotnet/runtime/blob/be98e88c760526452df94ef452fff4602fb5bded/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs#L113-L152>

Noticed while looking at <https://github.com/dotnet/runtime/pull/68022>.

**Comment by soroshsabz** [comment]
Did you want to resolve this, in my PR? or you want to create another PR?

**Comment by KalleOlaviNiemitalo** [comment]
I don't want to create a PR.

**Comment by tommcdon** [comment]
@tarekgh @noahfalk

**Comment by KalleOlaviNiemitalo** [comment]
Alternatively, it could be changed to throw ArgumentNullException, but then the parameter should not be marked as nullable.

**Comment by soroshsabz** [comment]
> Alternatively, it could be changed to throw ArgumentNullException, but then the parameter should not be marked as nullable.

@KalleOlaviNiemitalo I think this decision break API, so we have to handle nullability

**Comment by soroshsabz** [comment]
@tommcdon If you agree that this issue is bug, So I think, we have to resolve this for `.NET 6`

Are you agree with me?

**Comment by tarekgh** [comment]
I am going to fix this as originally suggested by @KalleOlaviNiemitalo in the description. The fix will be in .NET 7.0. I don't think we need to port this in .NET 6.0 as the workaround for users is they can check null in their code before calling. The users can recycle the listener too if needed.

**Comment by soroshsabz** [comment]
@tarekgh I add PR for this issue, please see and review it #68071 instead of create another one

**Comment by tarekgh** [comment]
I have replied on the issue. please have the discussion there.

**Comment by soroshsabz** [comment]
@tarekgh  I wrote my comment on the issue, is there any wrong?

**Comment by tarekgh** [comment]
> I wrote my comment on the issue, is there any wrong?

Sorry, I commented by mistake here 😄 Thanks for submitting the PR. I have added a comment on the PR.

**Comment by soroshsabz** [comment]
> I am going to fix this as originally suggested by @KalleOlaviNiemitalo in the description. The fix will be in .NET 7.0. I don't think we need to port this in .NET 6.0 as the workaround for users is they can check null in their code before calling. The users can recycle the listener too if needed.

@tarekgh I think, because .NET 6 is LTS and this bug is not change any API, so we can release this for .NET 6, without any cost. and demonstrates Microsoft 's commitment to support LTS .NET

**Comment by tarekgh** [comment]
> I think, because .NET 6 is LTS and this bug is not change any API, so we can release this for .NET 6, without any cost. and demonstrates Microsoft 's commitment to support LTS .NET

Usually, we do the servicing when we see any customer/user is blocked by that. I already mentioned some trivial workarounds if anyone wants to avoid this issue. Let's wait to see if more people run into this issue and then we can consider porting it. Thanks for your feedback.
